erts/emulator/beam/erl_bif_persistent.c

```diff
@@ -0,0 +1,816 @@
+         */
+        erts_atomic_inc_nob(&hash_table->refc);
+        BIF_TRAP2(&persistent_term_info_export, BIF_P, make_boxed(hp), res);
+    }
+}
+
+Uint
+erts_persistent_term_count(void)
+{
+    HashTable* hash_table = (HashTable *) erts_atomic_read_nob(&the_hash_table);
+    return hash_table->num_entries;
+}
+
+void
+erts_init_persistent_dumping(void)
+{
+    HashTable* hash_table = (HashTable *) erts_atomic_read_nob(&the_hash_table);
+    ErtsLiteralArea** area_p;
+    Uint i;
+
+    /*
+     * Overwrite the array of Eterms in the current hash table
+     * with pointers to literal areas.
+     */
+
+    erts_persistent_areas = (ErtsLiteralArea **) hash_table->term;
+    erts_num_persistent_areas = hash_table->num_entries;
+    area_p = erts_persistent_areas;
+    for (i = 0; i < hash_table->allocated; i++) {
+        Eterm term = hash_table->term[i];
+
+        if (is_boxed(term)) {
+            *area_p++ = term_to_area(term);
+        }
+    }
+}
+
+/*
+ * Local functions.
+ */
+
+static HashTable*
+create_initial_table(void)
+{
+    HashTable* hash_table;
+    int i;
+
+    hash_table = (HashTable *) erts_alloc(ERTS_ALC_T_PERSISTENT_TERM,
+                                          sizeof(HashTable)+sizeof(Eterm) *
+                                          (INITIAL_SIZE-1));
+    hash_table->allocated = INITIAL_SIZE;
+    hash_table->num_entries = 0;
+    hash_table->mask = INITIAL_SIZE-1;
+    hash_table->first_to_delete = 0;
+    hash_table->num_to_delete = 0;
+    erts_atomic_init_nob(&hash_table->refc, (erts_aint_t)1);
+    for (i = 0; i < INITIAL_SIZE; i++) {
+        hash_table->term[i] = NIL;
+    }
+    return hash_table;
+}
+
+static BIF_RETTYPE
+persistent_term_get_all_trap(BIF_ALIST_2)
+{
+    TrapData* trap_data = (TrapData *) BIF_ARG_1;
+    Eterm res = BIF_ARG_2;
+    Uint bump_reds;
+
+    ASSERT(is_big(BIF_ARG_1));
+    ASSERT(is_list(BIF_ARG_2));
+    trap_data = (TrapData *) boxed_val(BIF_ARG_1);
+    bump_reds = trap_data->remaining;
+    res = do_get_all(BIF_P, trap_data, res);
+    ASSERT(is_list(res));
+    if (trap_data->remaining > 0) {
+        BUMP_ALL_REDS(BIF_P);
+        BIF_TRAP2(&persistent_term_get_all_export, BIF_P, BIF_ARG_1, res);
+    } else {
+        /*
+         * Decrement ref count (and possibly delete the hash table
+         * and associated literal area).
+         */
+        dec_table_refc(BIF_P, trap_data->table);
+        BUMP_REDS(BIF_P, bump_reds);
+        BIF_RET(res);
+    }
+}
+
+static Eterm
+do_get_all(Process* c_p, TrapData* trap_data, Eterm res)
+{
+    HashTable* hash_table;
+    Uint remaining;
+    Uint idx;
+    Uint max_iter;
+    Uint i;
+    Eterm* hp;
+    Uint heap_size;
+    struct copy_term {
+        Uint key_size;
+        Eterm* tuple_ptr;
+    } *copy_data;
+
+    hash_table = trap_data->table;
+    idx = trap_data->idx;
+#if defined(DEBUG) || defined(VALGRIND)
+    max_iter = 50;
+#else
+    max_iter = c_p->fcalls;
```

Comment:
I just realized that you have to use `ERTS_BIF_REDS_LEFT`, otherwise you will get very strange results when save_calls is enabled.

Reply:
Fixed.